src/lib/ingest.js

```diff
@@ -1,53 +1,264 @@
+      // pass
+    } else {
+      error = JSON.stringify(error)
+    }
+    logger.error(`Error ingesting ${item.id}: ${error}`)
+    if (this.config.errorHandler) {
+      try {
+        await this.config.errorHandler(item, error)
+      } catch (err) {
+        logger.error(`Ingest error hander failed: ${err}`)
+      }
+    }
+  }
+
+  async ingestCollection(collection) {
+    try {
+      const index = COLLECTIONS_INDEX
+      const id = collection.id
+      const body = this.collectionToEs(collection)
+      await this.client.update({
+        index,
+        type: '_doc',
+        id,
+        body
+      })
+
+      logger.info(`Ingested collection ${id}; attempting to create index`)
+      await esClient.createIndex(id)
+      logger.info(`Index created for collection ${id}`)
+      this.collections.add(id)
+      await this.successHandler(this.addLinks(collection, body.doc))
+    } catch (error) {
+      await this.errorHandler(collection, error)
+    }
+  }
+
+  async ingestItems(items) {
+    const promises = []
+
+    // prepare items for ES query
+    const itemsById = {}
+    const operations = []
+    items.forEach((item) => {
+      try {
+        const { operation, body } = this.itemToEs(item)
+        itemsById[item.id] = {
+          original: item,
+          converted: body.doc
+        }
+        operations.push(operation)
+        operations.push(body)
+      } catch (err) {
+        promises.push(this.errorHandler(item, err))
+      }
+    })
+
+    if (operations.length === 0) {
+      logger.warn('No items remaining to be ingested')
+      return
+    }
+
+    logger.debug(`Writing batch of documents size ${operations.length / 2}`)
+    // run bulk ES query with items
+    await this.client.bulk({ body: operations }).then(async (response) => {
+      const esRespItems = response.body.items
+      logger.debug(`Wrote batch of documents size ${operations.length / 2}`)
+
+      // process ES response, handling successes and errors
+      let success = 0
+      esRespItems.forEach((esRespItem) => {
+        const esItem = esRespItem[(Object.keys(esRespItem)[0])]
+        const { original, converted } = itemsById[esItem._id]
+        if (esItem.error) {
+          promises.push(this.errorHandler(original, esItem.error))
         } else {
-          logger.debug('Ingested item')
-          resolve(true)
+          success += 1
+          promises.push(this.successHandler(this.addLinks(original, converted)))
         }
+      })
+      await Promise.allSettled(promises).then(() => {
+        logger.info(`${success} of ${items.length} items successfully ingested`)
+      })
+    })
+  }
+
+  async ingest(records) {
+    const items = []
+    const promises = []
+    const collectionPromises = []
+    records.forEach((record) => {
+      if (stacUtils.isItem(record)) {
+        items.push(record)
+      } else if (stacUtils.isCollection(record)) {
+        // collections we ingest one-by-one because they
```

Comment:
Where is it being enforced that collections are ingested one-by-one? It looks like all of the collections are being ingested in parallel.

Reply:
When I say one-by-one I mean using individual requests to ES rather than a single bulk request like is used for items.